federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -406,34 +406,33 @@ func (s *ServiceController) deleteClusterService(clusterName string, cachedServi
 
 func (s *ServiceController) ensureClusterService(cachedService *cachedService, clusterName string, service *v1.Service, client *kubeclientset.Clientset) error {
 	var err error
-	var needUpdate bool
 	for i := 0; i < clientRetryCount; i++ {
 		svc, err := client.Core().Services(service.Namespace).Get(service.Name)
 		if err == nil {
 			// service exists
 			glog.V(5).Infof("Found service %s/%s from cluster %s", service.Namespace, service.Name, clusterName)
-			//reserve immutable fields
-			service.Spec.ClusterIP = svc.Spec.ClusterIP
+			// we tend to update the new spec and
+			// current service spec(might have random cluster ip and node port) in the cluster identical
+
+			// cluster ip is not specified, update the auto assigned one to new service spec
+			if service.Spec.ClusterIP == "" {
+				service.Spec.ClusterIP = svc.Spec.ClusterIP
```

Comment:
ClusterIP is immutable property, it can not be changed during service updated. so the logic here is if ClusterIP is specified, we will passed it directly to k8s cluster, and surely, if the new specified ClusterIP is changed, the update will fail(and I don't think users should update a federation clusterIP is it is assigned by kubernetes cluster apiserver and will be different in multiple clusters.). If it is not specified, we will retrieve the existing value and apply to new spec, the purpose is not changing it but to retrieve the current value and reapply to make sure other properties can be updated successfully.